File: CamelUp.py
import random

# Number of camels and initial positions
num_camels = 5
camel_positions = [0] * num_camels  # All camels start at position 0

# Track the order of camels in the race; the first camel in the list is the bottom one in the stack
camel_order = list(range(num_camels))

# Function to roll the die for a camel
def roll_dice():
    return random.choice([1, 2, 3])  # Assuming dice have values 1-3 for simplicity

# Function to move camels based on dice roll
def move_camel(camel_index):
    roll = roll_dice()
    camel = camel_order[camel_index]
    current_position = camel_positions[camel]

    # Find all camels on top of the current camel and move them together
    stack_index = camel_index
    while stack_index < len(camel_order) and camel_positions[camel_order[stack_index]] == current_position:
        camel_positions[camel_order[stack_index]] += roll
        stack_index += 1

    # If camels reach a position with other camels, stack them together
    new_position = camel_positions[camel]
    if new_position in camel_positions[:camel_index]:
        # Find the bottom camel of the new stack and insert the moving camels above it
        bottom_camel_index = camel_positions.index(new_position)
        camel_order[bottom_camel_index+1:bottom_camel_index+1] = camel_order[camel_index:stack_index]

        # Remove the original positions of the moving camels
        del camel_order[camel_index:stack_index]

# Function to check for a winner
def check_winner():
    return any(position >= 16 for position in camel_positions)  # Assuming the track is 16 spaces long

# Function to print camel positions and stacks
def print_camel_info():
    print("Camel positions and order (bottom to top of stack):")
    for position in sorted(set(camel_positions)):
        stack = [camel for camel, pos in enumerate(camel_positions) if pos == position]
        print(f"Position {position}: {stack}")

# Main game loop
winner = False
turn_count = 0

while not winner:
    turn_count += 1
    print(f"\nTurn: {turn_count}")

    # Determine the move order based on the bottom-most camel in each stack
    move_order = sorted(range(num_camels), key=lambda x: (camel_positions[x], -x), reverse=True)

    for camel_index in move_order:
        move_camel(camel_index)
        print_camel_info()

        if check_winner():
            winner = True
            winning_camel = max(range(num_camels), key=lambda x: camel_positions[x])
            print(f"Game Over! Camel {winning_camel} wins!")
            break

    if not winner:
        input("Press Enter to continue to the next turn...")

File: Camel_Up.R
# Number of camels
num_camels <- 5

# Initialize camel positions at the start of the track
camel_positions <- rep(1, num_camels)

# Initialize stacks with each camel in its own stack
camel_stacks <- lapply(1:num_camels, function(x) x)

# Roll dice for one camel
roll_dice <- function() {
  sample(1:3, 1)  # Assuming dice have values 1-3 for simplicity
}

# Move camels
move_camel <- function(camel, dice_roll) {
  # Find the current position and stack of the camel
  current_position <- camel_positions[camel]
  stack_index <- which(sapply(camel_stacks, function(stack) camel %in% stack))
  
  # Move the camel and all camels on top of it
  move_stack <- camel_stacks[[stack_index]]
  camel_positions[move_stack] <- camel_positions[move_stack] + dice_roll
  
  # Check if moving to a stack with other camels
  if (any(camel_positions == current_position + dice_roll)) {
    target_stack_index <- which(sapply(camel_stacks, function(stack) any(camel_positions[stack] == current_position + dice_roll)))
    camel_stacks[[target_stack_index]] <- c(camel_stacks[[target_stack_index]], move_stack)
    camel_stacks[[stack_index]] <- NULL  # Remove the old stack
  } else {
    camel_stacks[[stack_index]] <- move_stack
  }
  
  # Remove any NULL entries in camel_stacks
  camel_stacks <- camel_stacks[sapply(camel_stacks, function(x) !is.null(x))]
}

# Check for winners
check_winner <- function() {
  if (any(camel_positions >= 16)) {  # Assuming the track is 16 spaces long
    return(TRUE)
  }
  return(FALSE)
}

# Print camel positions and stacks
print_camel_info <- function() {
  cat("Camel positions:\n")
  for (camel in 1:num_camels) {
    cat(sprintf("Camel %d is at position %d\n", camel, camel_positions[camel]))
  }
  cat("\nCamel stacks:\n")
  for (i in seq_along(camel_stacks)) {
    cat(sprintf("Stack %d contains camels: %s\n", i, paste(camel_stacks[[i]], collapse = ", ")))
  }
}

# Main game loop
winner <- FALSE
turn_count <- 0
remaining_camels <- 1:num_camels  # All camels start as 'not moved yet'

repeat {
  turn_count <- turn_count + 1
  cat("\nTurn:", turn_count, "\n")
  
  # Reset remaining camels for the next turn
  remaining_camels <- 1:num_camels
  
  while (length(remaining_camels) > 0) {
    # Select a random camel from those that haven't moved yet
    camel <- sample(remaining_camels, 1)
    
    # Roll the dice for the selected camel
    dice_roll <- roll_dice()
    
    # Move the selected camel
    move_camel(camel, dice_roll)
    
    # Remove the moved camel from the list of remaining camels
    remaining_camels <- remaining_camels[!remaining_camels %in% camel]
    
    # Print out current positions and stacks
    print_camel_info()
    
    # Check for a winner
    winner <- check_winner()
    
    if (winner) {
      cat("Game Over! We have a winner!\n")
      break
    }
  }
  
  if (winner) {
    break
  }
  
  # Wait for user input to proceed to the next turn
  readline(prompt = "Press [enter] to continue to the next turn...")
}
